timApp/tide_support/routes.py

```diff
@@ -0,0 +1,399 @@
+from timApp.bookmark.bookmarks import MY_COURSES_GROUP
+from timApp.document.docentry import DocEntry
+from timApp.document.docinfo import DocInfo
+from timApp.document.usercontext import UserContext
+from timApp.document.viewcontext import default_view_ctx, OriginInfo
+from timApp.plugin.containerLink import render_plugin_multi
+from timApp.plugin.plugin import Plugin, PluginRenderOptions, PluginWrap
+from timApp.plugin.pluginOutputFormat import PluginOutputFormat
+from timApp.plugin.taskid import TaskId
+from timApp.printing.printsettings import PrintFormat
+from timApp.user.user import User
+from tim_common.marshmallow_dataclass import class_schema
+
+
+@dataclass
+class IdeFile:
+    """
+    File that contains the code and path for one TIDE-task
+    """
+
+    by: str | None = None
+    """ Code of the file when plugin has only one file """
+
+    byCode: str | None = None
+    """ Code of the file when plugin has multiple files """
+
+    path: str | None = None
+    """ Path of the file for folder structure """
+
+    # Convert to json and set code based on 'by' or 'byCode'
+    def to_json(self):
+        return {
+            "code": self.by or self.byCode,
+            "path": self.path,
+        }
+
+
+IdeFileSchema = class_schema(IdeFile)()
+
+
+@dataclass
+class TIDETaskInfo:
+    """
+    Information about the TIDE-task
+    """
+
+    header: str | None = None
+    """
+    Header of the plugin
+    """
+
+    stem: str | None = None
+    """
+    Stem of the plugin
+    """
+
+    answer_count: int | None = None
+    """
+    Number of answers for the task
+    """
+
+    # Might be needed to require this
+    type: str | None = None
+    """ Type of the file """
+
+
+TIDETaskInfoSchema = class_schema(TIDETaskInfo)()
+
+
+def user_ide_courses(user: User) -> list[json]:
+    """
+    Gets all courses that have parameter for Ide course in course settings and are bookmarked by the user
+    :param user: Logged-in user
+    :return: List JSON with all TIDE-tasks from the courses user has bookmarked
+    """
+    user_courses = user.bookmarks.bookmark_data[2][MY_COURSES_GROUP]
+    ide_courses = []
+
+    for course_dict in user_courses:
+        for course_name, course_path in course_dict.items():
+            # Get the document by path, remove /view/ from the beginning of path
+            doc = DocInfo.find_by_path(course_path.split("view/")[1:][0])
+
+            if doc is None:
+                continue
+
+            demo_paths = doc.document.get_settings().ide_course()
+
+            if demo_paths is None:
+                continue
+
+            ide_courses.append(
+                {
+                    "course_name": course_name,
+                    "course_id": doc.document.doc_id,
+                    "course_path": course_path,
+                    "demo_paths": demo_paths,
+                }
+            )
+
+    return ide_courses
+
+
+def demos_by_doc(doc_id: int = None, doc_path: str = None):
```

Comment:
Tyypitys: None pitää merkitä tyyppiin erikseen, esim `int | None`.

```suggestion
def demos_by_doc(doc_id: int | None = None, doc_path: str | None = None) -> list[IdeDocument]:
```